### PR TITLE
Implemented completion for dictionary items

### PR DESCRIPTION
**Pull request by mardukbp**
Dictionary items can now be completed using the syntax

```
${dict}[some index][another index]
```

That is, dictionary items, which are themselves dictionaries are also handled.

The code is meant to be short and simple. Feel free to adapt it as you see fit.

### CODE REVIEW

**Review by fabioz — CHANGES_REQUESTED** [review]
Thanks for the pull request.

I made some general comments which I believe are straightforward to address.

Also, this pull request needs some tests along with it.

You can add create a new `test_dictionary_variables_completions.py` for this feature based on:
https://github.com/robocorp/robotframework-lsp/blob/master/robotframework-ls/tests/robotframework_ls_tests/completions/test_variables_completions.py.

File: robotframework-ls/src/robotframework_ls/impl/dictionary_completions.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+                                  CompletionItemKind,
+                                  InsertTextFormat,
+                                  Position,
+                                  Range,
+                                  TextEdit)
+
+log = get_logger(__name__)
+
+
+def get_variables(completion_context):
```
[the remark this answers is not in the document]
Comment (mardukbp) [reply robotframework-ls/src/robotframework_ls/impl/dictionary_completions.py]: Done

File: robotframework-ls/src/robotframework_ls/impl/dictionary_completions.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    variables = []
+    ast = completion_context.get_ast()
+    for node_info in ast_utils.iter_variables(ast):
```
[the remark this answers is not in the document]
Comment (mardukbp) [reply robotframework-ls/src/robotframework_ls/impl/dictionary_completions.py]: Done. I think it is better to refactor variable_completions to use CompletionContext.get_all_variables after merging this PR.

File: robotframework-ls/src/robotframework_ls/impl/dictionary_completions.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        node = node_info.node
+        var_name = node.get_token(Token.VARIABLE).value
```
[the remark this answers is not in the document]
Comment (mardukbp) [reply robotframework-ls/src/robotframework_ls/impl/dictionary_completions.py]: Done.

File: robotframework-ls/src/robotframework_ls/impl/dictionary_completions.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        var_value = node.value
+        variables.append((var_name, var_value))
+    return variables
+
+
+def get_dict_name(dict_item_access):
+    dict_variable_regex = re.compile(r"[$|&]{(\w+)}")
```
[the remark this answers is not in the document]
Comment (mardukbp) [reply robotframework-ls/src/robotframework_ls/impl/dictionary_completions.py]: Done.

File: robotframework-ls/src/robotframework_ls/impl/dictionary_completions.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    ast = completion_context.get_ast()
+    for node_info in ast_utils.iter_variables(ast):
+        node = node_info.node
+        var_name = node.get_token(Token.VARIABLE).value
+        var_value = node.value
+        variables.append((var_name, var_value))
+    return variables
+
+
+def get_dict_name(dict_item_access):
```
[the remark this answers is not in the document]
Comment (mardukbp) [reply robotframework-ls/src/robotframework_ls/impl/dictionary_completions.py]: Done.

File: robotframework-ls/src/robotframework_ls/impl/dictionary_completions.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    dict_variable_regex = re.compile(r"[$|&]{(\w+)}")
+    dict_name = dict_variable_regex.findall(dict_item_access)
+    return "&{" + dict_name.pop() + "}"
+
+
+def get_dict_keys(dict_item_access):
```
[the remark this answers is not in the document]
Comment (mardukbp) [reply robotframework-ls/src/robotframework_ls/impl/dictionary_completions.py]: Done.

File: robotframework-ls/src/robotframework_ls/impl/dictionary_completions.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    dict_key_regex = re.compile(r"\[([\w\s]*)\]")
```
[the remark this answers is not in the document]
Comment (mardukbp) [reply robotframework-ls/src/robotframework_ls/impl/dictionary_completions.py]: Done.

File: robotframework-ls/src/robotframework_ls/impl/dictionary_completions.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    dict_keys = dict_key_regex.findall(dict_item_access)
+    dict_keys.reverse()
+    return dict_keys
+
+
+def get_dictionary(variables, dict_name, dict_keys):
```
[the remark this answers is not in the document]
Comment (mardukbp) [reply robotframework-ls/src/robotframework_ls/impl/dictionary_completions.py]: Done.

File: robotframework-ls/src/robotframework_ls/impl/dictionary_completions.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+                return dictionary
+            else:
+                dict_variable = dictionary[dict_entry]
+                dict_name = get_dict_name(dict_variable)
+                dict_keys += get_dict_keys(dict_variable)
+                return get_dictionary(variables, dict_name, dict_keys)
+    return None
+
+
+def as_dictionary(dict_tokens):
```
[the remark this answers is not in the document]
Comment (mardukbp) [reply robotframework-ls/src/robotframework_ls/impl/dictionary_completions.py]: Done.

File: robotframework-ls/src/robotframework_ls/impl/dictionary_completions.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    """
+    Parse ["key1=val1", "key2=val2",...] as a dictionary
+    """
+    return dict([token.split("=") for token in dict_tokens])
+
+
+def completion_items(dictionary, editor_range):
```
[the remark this answers is not in the document]
Comment (mardukbp) [reply robotframework-ls/src/robotframework_ls/impl/dictionary_completions.py]: Done.

File: robotframework-ls/src/robotframework_ls/impl/dictionary_completions.py [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+             kind=CompletionItemKind.Variable,
+             text_edit=TextEdit(editor_range, f"{key}]"),
+             insertText=key,
+             documentation=value,
+             insertTextFormat=InsertTextFormat.Snippet,
+             documentationFormat=MarkupKind.PlainText,
+             ).to_dict() for key, value in dictionary.items()]
+
+
+def complete(completion_context):
```
[the remark this answers is not in the document]
Comment (mardukbp) [reply robotframework-ls/src/robotframework_ls/impl/dictionary_completions.py]: Done.

File: robotframework-ls/tests/robotframework_ls_tests/completions/test_dictionary_entries_completions.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+&{Person}   First name=John   Last name=Smith
+
+*** Test Cases ***
+Dictionary Variable
+    Log to Console    ${Person}[First]"""
+    line, col = doc.get_last_line_col()
+    completions = dictionary_completions.complete(
+        CompletionContext(doc, workspace=workspace.ws, line=line, col=col - len("]"))
+    )
+    data_regression.check(completions)
```
Comment (fabioz) [inline robotframework-ls/tests/robotframework_ls_tests/completions/test_dictionary_entries_completions.py:20]: The actual files related to the data regression aren't added (so, tests will fail in the ci).

**Comment by fabioz** [comment]
Just merged. Thanks for the pull request.

**Pull request merged**